src/type.js

```diff
@@ -710,29 +384,219 @@ Sk.builtin.type.buildMRO_ = function (klass) {
+    },
+    __name__: {
+        $get: function () {
+            return new Sk.builtin.str(this.prototype.tp$name);
+        },
+        $set: function (value) {
+            if (!Sk.builtin.checkString(value)) {
+                throw new Sk.builtin.TypeError(
+                    "can only assign string to " + this.prototype.tp$name + ".__name__, not '" + Sk.abstr.typeName(value) + "'"
+                );
+            }
+            this.prototype.tp$name = value.$jsstr();
+        },
+    },
+    __module__: {
+        $get: function () {
+            let mod = this.prototype.__module__;
+            if (mod && !(mod.ob$type === Sk.builtin.getset_descriptor)) {
+                return mod;
+            }
+            return new Sk.builtin.str("builtins");
+        },
+        $set: function (value) {
+            // they can set the module to whatever they like
+            this.prototype.__module__ = value;
+        },
+    },
 };
 
-Sk.builtin.type.pythonFunctions = ["__format__"];
+Sk.builtin.type.prototype.tp$methods = /**@lends {Sk.builtin.type.prototype}*/ {
+    mro: {
+        $meth: function () {
+            return new Sk.builtin.tuple(this.$buildMRO());
+        },
+        $flags: { NoArgs: true },
+    },
+    __dir__: {
+        $meth: function __dir__() {
+            const seen = new Set();
+            const dir = [];
+            function push_or_continue(attr) {
+                if (attr in Sk.reservedWords_) {
+                    return;
+                }
+                attr = Sk.unfixReserved(attr);
+                if (attr.indexOf("$") !== -1) {
+                    return;
+                }
+                if (!seen.has(attr)) {
+                    seen.add(attr);
+                    dir.push(new Sk.builtin.str(attr));
+                }
+            }
+            if (this.prototype.sk$prototypical) {
+                for (let attr in this.prototype) {
+                    push_or_continue(attr);
+                }
+            } else {
+                const mro = this.prototype.tp$mro;
+                for (let i = 0; i < mro.length; i++) {
+                    const attrs = Object.getOwnPropertyNames(mro[i].prototype);
+                    for (let j = 0; j < attrs.length; j++) {
+                        push_or_continue(attrs[j]);
+                    }
+                }
+            }
+            return new Sk.builtin.list(dir);
+        },
+        $flags: { NoArgs: true },
+        $doc: "Specialized __dir__ implementation for types.",
+    },
+};
+
+// we could move this to the prototype but this is called before the klass constructor inheritance is set
+// this function is used to determine the class constructor inheritance.
+Sk.builtin.type.$best_base = function (bases) {
+    if (bases.length === 0) {
+        bases.push(Sk.builtin.object);
+    }
+    function solid_base(type) {
+        // if we support slots we would need to change this function - for now it just checks for the builtin.
```

Comment:
Could you explain this? You're pretty far along implementing slots right?

Reply:
Yes. So weirdly the best base is not always the first bases in the tuple of bases! Who knew...

Eg
class A: pass
class B(A, int): pass
B.__base__ = int

This algorithm for finding best base follows a simplified version of the cpython algorithm.

Currently the solid base is just the first builtin we find. The solid base is what is required to check for inheritance conflicts. 

When slots are in the game then the solid base will be something like the first class we find with slots. 

There are certain rules about inheriting from classes with slots that do and don’t cause layout conflicts... this algorithm would deal with those layout conflicts eg. 

class A: __slots__= ‘foo’
class B: __slots__= ‘bar’
class C(A,B): pass # error layout conflict 

For now the only layout conflicts we can have are conflicts from inheriting from multiple builtins. 

class A(dict): pass
class B(int): pass
class C(A,B): pass # layout conflicts the solid bases of A and B are not compatible.

Reply:
Here are the cpython versions 


https://github.com/python/cpython/blob/4a21e57fe55076c77b0ee454e1994ca544d09dc0/Objects/typeobject.c#L2091


https://github.com/python/cpython/blob/4a21e57fe55076c77b0ee454e1994ca544d09dc0/Objects/typeobject.c#L2012

Reply:
ps I think I unwittingly hit resolve on a few of these by mistake on my phone. And hence the unformatted code snippets! Sorry.

Reply:
python is always weirder then you expect, when you look at it as closely as we do when making skulpt. 😄 this is such a good example of that